File: src/main/java/com/ecommerce/inventory/event/BaseEvent.java
package com.ecommerce.inventory.event;

import org.springframework.context.ApplicationEvent;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Base class for all domain events in the system.
 * Provides common event metadata and versioning support.
 */
public abstract class BaseEvent extends ApplicationEvent {
    
    private final String eventId;
    private final LocalDateTime timestamp;
    private final String eventVersion;
    private final String eventType;
    private final String aggregateId;
    private final String userId;
    
    protected BaseEvent(Object source, String aggregateId, String userId) {
        super(source);
        this.eventId = UUID.randomUUID().toString();
        this.timestamp = LocalDateTime.now();
        this.eventVersion = "1.0";
        this.eventType = this.getClass().getSimpleName();
        this.aggregateId = aggregateId;
        this.userId = userId;
    }
    
    protected BaseEvent(Object source, String aggregateId, String userId, String eventVersion) {
        super(source);
        this.eventId = UUID.randomUUID().toString();
        this.timestamp = LocalDateTime.now();
        this.eventVersion = eventVersion;
        this.eventType = this.getClass().getSimpleName();
        this.aggregateId = aggregateId;
        this.userId = userId;
    }
    
    public String getEventId() {
        return eventId;
    }
    
    public LocalDateTime getTimestamp() {
        return timestamp;
    }
    
    public String getEventVersion() {
        return eventVersion;
    }
    
    public String getEventType() {
        return eventType;
    }
    
    public String getAggregateId() {
        return aggregateId;
    }
    
    public String getUserId() {
        return userId;
    }
    
    @Override
    public String toString() {
        return String.format("%s{eventId='%s', timestamp=%s, version='%s', aggregateId='%s', userId='%s'}", 
                eventType, eventId, timestamp, eventVersion, aggregateId, userId);
    }
}

File: apps/api/src/events/types/BaseEvent.ts
/**
 * Base interface for all domain events in the system.
 * Provides common event metadata and versioning support.
 */
export interface BaseEvent {
  eventId: string;
  timestamp: Date;
  eventVersion: string;
  eventType: string;
  aggregateId: string;
  userId: string;
}

/**
 * Abstract base class for creating domain events.
 */
export abstract class DomainEvent implements BaseEvent {
  public readonly eventId: string;
  public readonly timestamp: Date;
  public readonly eventVersion: string;
  public readonly eventType: string;
  public readonly aggregateId: string;
  public readonly userId: string;

  protected constructor(aggregateId: string, userId: string, eventVersion: string = '1.0') {
    this.eventId = crypto.randomUUID();
    this.timestamp = new Date();
    this.eventVersion = eventVersion;
    this.eventType = this.constructor.name;
    this.aggregateId = aggregateId;
    this.userId = userId;
  }

  toString(): string {
    return `${this.eventType}{eventId='${this.eventId}', timestamp=${this.timestamp.toISOString()}, version='${this.eventVersion}', aggregateId='${this.aggregateId}', userId='${this.userId}'}`;
  }
}

/**
 * Event metadata for tracking and monitoring.
 */
export interface EventMetadata {
  eventId: string;
  eventType: string;
  aggregateId: string;
  userId: string;
  timestamp: Date;
  version: string;
  correlationId?: string;
  causationId?: string;
  source?: string;
}

/**
 * Event processing result.
 */
export interface EventProcessingResult {
  success: boolean;
  eventId: string;
  eventType: string;
  processingTime: number;
  error?: string;
  retryCount?: number;
}